File: mobile_ios_lib/MaplatView/MaplatView.m
//
//  MaplatView.m
//  MaplatView
//
//  Created by Takashi Irie on 2018/07/03.
//  Copyright © 2018 TileMapJp. All rights reserved.
//

#import "MaplatView.h"
#import "MaplatCache.h"

@interface MaplatView () <MaplatCacheDelegate>

@property (nonatomic, strong) MaplatCache *cache;
@property (nonatomic, strong) NSDictionary *initializeValue;

@end

@implementation MaplatView

+ (void)configure {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    NSString *path = [[paths objectAtIndex:0] stringByAppendingString:@"/webCache"];
    
    NSURLCache *defaultCache = [NSURLCache sharedURLCache];
    NSURLCache *maplatCache = [[MaplatCache alloc] initWithMemoryCapacity:defaultCache.memoryCapacity
                                                             diskCapacity:defaultCache.diskCapacity
                                                                 diskPath:path];
    [NSURLCache setSharedURLCache:maplatCache];
}

- (instancetype)initWithFrame:(CGRect)frame appID:(NSString *)appID setting:(NSDictionary *)setting {
    self = [super initWithFrame:frame];
    if (!self) return nil;
    
    _cache = (MaplatCache *)[NSURLCache sharedURLCache];
    _cache.delegate = self;
    
    _webView = [[UIWebView alloc] initWithFrame:self.bounds];
    _webView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _webView.delegate = _cache;
    [self addSubview:_webView];
    
    [_webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"http://localresource/mobile.html"]]];
    
    if (!appID) {
        appID = @"mobile";
    }
    NSMutableDictionary *jsonObj = [NSMutableDictionary new];
    [jsonObj setValue:appID forKey:@"appid"];
    if (setting) {
        [jsonObj setValue:setting forKey:@"setting"];
    }
    _initializeValue = jsonObj;
    
    return self;
}

#pragma mark - MaplatCacheDelegate

- (void)onCallWeb2AppWithKey:(NSString *)key value:(NSString *)value {
    if (!_delegate) return;
    
    NSLog(@"onCallWeb2AppWithKey:%@ value:%@", key, value);
    if ([key isEqualToString:@"ready"]) {
        if ([value isEqualToString:@"callApp2Web"]) {
            [self callApp2WebWithKey:@"maplatInitialize" value:_initializeValue];
        } else if ([value isEqualToString:@"maplatObject"]) {
            [_delegate onReady];
        }
    } else if ([key isEqualToString:@"clickMarker"]) {
        NSData *jsonData = [value dataUsingEncoding:NSUTF8StringEncoding];
        NSDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:jsonData
                                                                   options:kNilOptions
                                                                     error:nil];
        long markerId = [(NSNumber *)jsonObject[@"id"] longValue];
        id markerData = jsonObject[@"data"];
        [_delegate onClickMarkerWithMarkerId:markerId markerData:markerData];
    } else if ([key isEqualToString:@"changeViewpoint"]) {
        NSData *jsonData = [value dataUsingEncoding:NSUTF8StringEncoding];
        NSDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:jsonData
                                                                   options:kNilOptions
                                                                     error:nil];
        double x = [(NSNumber *)jsonObject[@"x"] doubleValue];
        double y = [(NSNumber *)jsonObject[@"y"] doubleValue];
        double longitude = [(NSNumber *)jsonObject[@"longitude"] doubleValue];
        double latitude = [(NSNumber *)jsonObject[@"latitude"] doubleValue];
        double mercator_x = [(NSNumber *)jsonObject[@"mercator_x"] doubleValue];
        double mercator_y = [(NSNumber *)jsonObject[@"mercator_y"] doubleValue];
        double zoom = [(NSNumber *)jsonObject[@"zoom"] doubleValue];
        double merc_zoom = [(NSNumber *)jsonObject[@"merc_zoom"] doubleValue];
        double direction = [(NSNumber *)jsonObject[@"direction"] doubleValue];
        double rotation = [(NSNumber *)jsonObject[@"rotation"] doubleValue];
        
        [_delegate onChangeViewpointWithX:x y:y latitude:latitude longitude:longitude mercatorX:mercator_x mercatorY:mercator_y zoom:zoom mercZoom:merc_zoom direction:direction rotation:rotation];
    } else if([key isEqualToString:@"outOfMap"]) {
        [_delegate onOutOfMap];
    } else if ([key isEqualToString:@"clickMap"]) {
        NSData *jsonData = [value dataUsingEncoding:NSUTF8StringEncoding];
        NSDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:jsonData
                                                                   options:kNilOptions
                                                                     error:nil];
        double longitude = [(NSNumber *)jsonObject[@"longitude"] doubleValue];
        double latitude = [(NSNumber *)jsonObject[@"latitude"] doubleValue];
        [_delegate onClickMapWithLatitude:latitude longitude:longitude];
    }
}

- (void)callApp2WebWithKey:(NSString *)key value:(id)value
{
    [self callApp2WebWithKey:key value:value callback:nil];
}

- (void)callApp2WebWithKey:(NSString *)key value:(id)value callback:(void (^)(NSString *))callback
{
    NSString* jsonStr;
    if (value == nil) {
        jsonStr = nil;
    } else if ([value isKindOfClass:[NSString class]]) {
        jsonStr = value;
    } else if ([value isKindOfClass:[NSNumber class]]) {
        jsonStr = [value stringValue];
    } else {
        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:value options:0 error:nil];
        jsonStr = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    }
    NSString *retVal = [_webView stringByEvaluatingJavaScriptFromString:[NSString stringWithFormat:@"maplatBridge.callApp2Web('%@','%@');", key, jsonStr]];
    if (callback != nil) {
        callback(retVal);
    }
}

- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId stringData:(NSString *)markerData
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId stringData:markerData iconUrl:nil];
}
- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId stringData:(NSString *)markerData iconUrl:(NSString *) iconUrl
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId markerData:markerData iconUrl:iconUrl];
}
- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId longData:(long)markerData
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId longData:markerData iconUrl:nil];
}
- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId longData:(long)markerData iconUrl:(NSString *) iconUrl
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId markerData:[NSNumber numberWithLong:markerData] iconUrl:iconUrl];
}
- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId doubleData:(double)markerData
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId doubleData:markerData iconUrl:nil];
}
- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId doubleData:(double)markerData iconUrl:(NSString *) iconUrl
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId markerData:[NSNumber numberWithDouble:markerData] iconUrl:iconUrl];
}
- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId arrayData:(NSArray *)markerData
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId arrayData:markerData iconUrl:nil];
}
- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId arrayData:(NSArray *)markerData iconUrl:(NSString *) iconUrl
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId markerData:markerData iconUrl:iconUrl];
}
- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId dictData:(NSDictionary *)markerData
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId dictData:markerData iconUrl:nil];
}
- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId dictData:(NSDictionary *)markerData iconUrl:(NSString *) iconUrl
{
    [self addMarkerWithLatitude:latitude longitude:longitude markerId:markerId markerData:markerData iconUrl:iconUrl];
}

- (void)addMarkerWithLatitude:(double)latitude longitude:(double)longitude markerId:(long)markerId markerData:(id)markerData iconUrl:(NSString *)iconUrl
{
    NSMutableDictionary *jsonObj = [NSMutableDictionary new];
    NSArray *Lnglat = @[[NSNumber numberWithDouble:longitude], [NSNumber numberWithDouble:latitude]];
    [jsonObj setValue:Lnglat forKey:@"lnglat"];
    [jsonObj setValue:[NSNumber numberWithLong:markerId] forKey:@"id"];
    [jsonObj setValue:markerData forKey:@"data"];
    if ([iconUrl length] > 0) {
        [jsonObj setValue:iconUrl forKey:@"icon"];
    }
    [self callApp2WebWithKey:@"addMarker" value:jsonObj];
}

- (void)clearMarker
{
    [self callApp2WebWithKey:@"clearMarker" value:nil];
}

- (void)setGPSMarkerWithLatitude:(double)latitude longitude:(double)longitude accuracy:(double)accuracy
{
    NSMutableDictionary *jsonObj = [NSMutableDictionary new];
    NSArray *Lnglat = @[[NSNumber numberWithDouble:longitude], [NSNumber numberWithDouble:latitude]];
    [jsonObj setValue:Lnglat forKey:@"lnglat"];
    [jsonObj setValue:[NSNumber numberWithDouble:accuracy] forKey:@"acc"];
    [self callApp2WebWithKey:@"setGPSMarker" value:jsonObj];
}

- (void)changeMap:(NSString *)mapID
{
    [self callApp2WebWithKey:@"changeMap" value:mapID];
}

- (void)setViewpointWithLatitude:(double)latitude longitude:(double)longitude
{
    NSMutableDictionary *jsonObj = [NSMutableDictionary new];
    [jsonObj setValue:[NSNumber numberWithDouble:longitude] forKey:@"longitude"];
    [jsonObj setValue:[NSNumber numberWithDouble:latitude] forKey:@"latitude"];
    [self callApp2WebWithKey:@"setViewpoint" value:jsonObj];
}

- (void)setDirection:(double)direction
{
    NSMutableDictionary *jsonObj = [NSMutableDictionary new];
    [jsonObj setValue:[NSNumber numberWithDouble:direction] forKey:@"direction"];
    [self callApp2WebWithKey:@"setViewpoint" value:jsonObj];
}

- (void)setRotation:(double)rotation
{
    NSMutableDictionary *jsonObj = [NSMutableDictionary new];
    [jsonObj setValue:[NSNumber numberWithDouble:rotation] forKey:@"rotation"];
    [self callApp2WebWithKey:@"setViewpoint" value:jsonObj];
}

- (void)addLineWithLngLat:(NSArray *)lnglats stroke:(NSDictionary *)stroke
{
    NSMutableDictionary *jsonObj = [NSMutableDictionary new];
    [jsonObj setValue:lnglats forKey:@"lnglats"];
    if (stroke) {
        [jsonObj setValue:stroke forKey:@"stroke"];
    }
    [self callApp2WebWithKey:@"addLine" value:jsonObj];
}

- (void)clearLine
{
    [self callApp2WebWithKey:@"clearLine" value:nil];
}

- (void)currentMapID:(void (^)(NSString *))callback
{
    [self callApp2WebWithKey:@"currentMapID" value:nil callback:callback];
}

- (void)currentMapInfo:(void (^)(NSDictionary *))callback
{
    [self callApp2WebWithKey:@"currentMapInfo" value:nil callback:^(NSString *value){
        if ([value isEqualToString:@""]) {
            callback(nil);
            return;
        }
        NSData *jsonData = [value dataUsingEncoding:NSUTF8StringEncoding];
        NSDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:jsonData
                                                                   options:kNilOptions
                                                                     error:nil];
        callback(jsonObject);
    }];
}

- (void)mapInfo:(NSString *)sourceID callback:(void (^)(NSDictionary *))callback
{
    [self callApp2WebWithKey:@"mapInfo" value:sourceID callback:^(NSString *value){
        if ([value isEqualToString:@""]) {
            callback(nil);
            return;
        }
        NSData *jsonData = [value dataUsingEncoding:NSUTF8StringEncoding];
        NSDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:jsonData
                                                                   options:kNilOptions
                                                                     error:nil];
        callback(jsonObject);
    }];
}

@end

File: mobile_ios/mobile_ios_lib/MaplatView/MaplatView.swift
//
//  MaplatView.m
//  MaplatView
//
//  Created by Takashi Irie on 2018/07/03.
//  Copyright © 2018 TileMapJp. All rights reserved.
//

import Foundation
import UIKit

public class MaplatView: UIView, MaplatCacheDelegate {
    
    var cache: MaplatCache?
    var initializeValue: [String : Any]?
    var webView: UIWebView?
    
    @objc
    public var delegate: MaplatViewDelegate?
    
    @objc
    class public func configure() {
        let paths = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)
        let path = paths[0].appending("/webCache")
        let defaultCache = URLCache.shared
        let maplatCache = MaplatCache(memoryCapacity: defaultCache.memoryCapacity, diskCapacity: defaultCache.diskCapacity, diskPath: path)
        URLCache.shared = maplatCache
    }
    
    @objc
    public init(frame: CGRect, appID: String?, setting: [String : Any]?) {
        super.init(frame: frame)
        self.cache = (URLCache.shared as! MaplatCache)
        self.cache!.delegate = self
        self.webView = UIWebView(frame: self.bounds)
        self.webView!.autoresizingMask = [.flexibleWidth,.flexibleHeight]
        self.webView!.delegate = self.cache
        self.addSubview(self.webView!)
        self.webView!.loadRequest(URLRequest(url: URL(string: "http://localresource/mobile.html")!))
        var jsonObj = [String : Any]()
        jsonObj["appid"] = appID == nil ? "mobile" : appID
        if (setting != nil) {
            jsonObj["setting"] = setting
        }
        self.initializeValue = jsonObj
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func toDictionary(_ value: String) -> [String: Any]? {
        if let data = value.data(using: .utf8) {
            do {
                return try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
            } catch {
                print(error.localizedDescription)
            }
        }
        return nil
    }
    
    // MARK: - MaplatCacheDelegate
    public func onCallWeb2App(_ key: String, value: String) {
        guard let delegate = self.delegate else {
            return
        }
        
        print("onCallWeb2App:\(key) value:\(value)")
        if (key == "ready") {
            if (value == "callApp2Web") {
                self.callApp2Web("maplatInitialize", value: [self.initializeValue as Any])
            } else if (value == "maplatObject") {
                delegate.onReady()
            }
        } else if (key == "clickMarker") {
            if let jsonObject = toDictionary(value) {
                if let markerId = jsonObject["id"] {
                    let markerData = jsonObject["data"]
                    delegate.onClickMarker(markerId: String(describing: markerId), markerData: markerData)
                }
            }
        } else if (key == "changeViewpoint") {
            if let jsonObject = toDictionary(value),
                let x = (jsonObject["x"] as? NSNumber)?.doubleValue,
                let y = (jsonObject["y"] as? NSNumber)?.doubleValue,
                let longitude = (jsonObject["longitude"] as? NSNumber)?.doubleValue,
                let latitude = (jsonObject["latitude"] as? NSNumber)?.doubleValue,
                let mercator_x = (jsonObject["mercator_x"] as? NSNumber)?.doubleValue,
                let mercator_y = (jsonObject["mercator_y"] as? NSNumber)?.doubleValue,
                let zoom = (jsonObject["zoom"] as? NSNumber)?.doubleValue,
                let merc_zoom = (jsonObject["merc_zoom"] as? NSNumber)?.doubleValue,
                let direction = (jsonObject["direction"] as? NSNumber)?.doubleValue,
                let rotation = (jsonObject["rotation"] as? NSNumber)?.doubleValue {
                delegate.onChangeViewpoint(x: x, y: y, latitude: latitude, longitude: longitude, mercatorX: mercator_x, mercatorY: mercator_y, zoom: zoom, mercZoom: merc_zoom, direction: direction, rotation: rotation)
            }
        } else if (key == "outOfMap") {
            delegate.onOutOfMap()
        } else if (key == "clickMap") {
            if let jsonObject = toDictionary(value) {
                if let longitude = (jsonObject["longitude"] as? NSNumber)?.doubleValue,
                    let latitude = (jsonObject["latitude"] as? NSNumber)?.doubleValue {
                    delegate.onClickMap(latitude: latitude, longitude: longitude)
                }
            }
        }
    }
    
    @objc
    public func callApp2Web(_ key: String, value: [Any]) {
        self.callApp2Web(key, value: value, callback: nil)
    }
    
    @objc
    public func callApp2Web(_ key: String, value: [Any], callback: ((String?) -> ())?) {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: value, options: [])
            let jsonStr = String(data: jsonData, encoding: .utf8)!
            let retVal = self.webView!.stringByEvaluatingJavaScript(from: "maplatBridge.callApp2Web('\(key)','\(jsonStr)');")
            if let callback = callback {
                callback(retVal)
            }
        } catch {
            print(error.localizedDescription)
        }
        
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, stringData markerData: String) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, stringData: markerData, iconUrl: nil)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, stringData markerData: String, iconUrl: String?) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, markerData: markerData, iconUrl: iconUrl)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, longData markerData: CLong) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, longData: markerData, iconUrl: nil)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, longData markerData: CLong, iconUrl: String?) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, markerData: NSNumber(value: markerData), iconUrl: iconUrl)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, doubleData markerData: CDouble) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, doubleData: markerData, iconUrl: nil)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, doubleData markerData: CDouble, iconUrl: String?) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, markerData: NSNumber(value: markerData), iconUrl: iconUrl)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, arrayData markerData: [Any]) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, arrayData: markerData, iconUrl: nil)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, arrayData markerData: [Any], iconUrl: String?) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, markerData: markerData, iconUrl: iconUrl)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, dictData markerData: [NSObject : Any]) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, dictData: markerData, iconUrl: nil)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, dictData markerData: [NSObject : Any], iconUrl: String?) {
        self.addMarkerWithLatitude(latitude, longitude: longitude, markerId: markerId, markerData: markerData, iconUrl: iconUrl)
    }
    
    @objc
    public func addMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, markerId: CLong, markerData: Any, iconUrl: String?) {
        var jsonObj: [String: Any] = [
            "lnglat": [NSNumber(value: longitude),NSNumber(value: latitude)],
            "id": NSNumber(value: markerId),
            "data": markerData
        ]
        if let iconUrl = iconUrl, iconUrl.count > 0 {
            jsonObj["icon"] = iconUrl
        }
        self.callApp2Web("addMarker", value: [jsonObj])
    }
    
    @objc
    public func clearMarker() {
        self.callApp2Web("clearMarker", value: [])
    }
    
    @objc
    public func setGPSMarkerWithLatitude(_ latitude: CDouble, longitude: CDouble, accuracy: CDouble) {
        let jsonObj: [String: Any] = [
            "lnglat": [NSNumber(value: longitude),NSNumber(value: latitude)],
            "acc": NSNumber(value: accuracy)
        ]
        self.callApp2Web("setGPSMarker", value: [jsonObj])
    }
    
    @objc
    public func changeMap(_ mapID: String) {
        self.callApp2Web("changeMap", value: [mapID])
    }
    
    @objc
    public func setViewpointWithLatitude(_ latitude: CDouble, longitude: CDouble) {
        let jsonObj = [
            "longitude": NSNumber(value: longitude),
            "latitude": NSNumber(value: latitude)
        ]
        self.callApp2Web("setViewpoint", value: [jsonObj])
    }
    
    @objc
    public func setDirection(_ direction: CDouble) {
        let jsonObj = ["direction": NSNumber(value: direction)]
        self.callApp2Web("setViewpoint", value: [jsonObj])
    }
    
    @objc
    public func setRotation(_ rotation: CDouble) {
        let jsonObj = ["rotation": NSNumber(value: rotation)]
        self.callApp2Web("setViewpoint", value: [jsonObj])
    }
    
    @objc
    public func addLineWithLngLat(_ lnglats: [Any], stroke: [String : Any]?) {
        var jsonObj: [String: Any] = [
            "lnglats": lnglats
        ]
        if let stroke = stroke {
            jsonObj["stroke"] = stroke
        }
        self.callApp2Web("addLine", value: [jsonObj])
    }
    
    @objc
    public func clearLine() {
        self.callApp2Web("clearLine", value: [])
    }
    
    @objc
    public func currentMapID(_ callback: @escaping (String?) -> ()) {
        self.callApp2Web("currentMapID", value: [], callback: callback)
    }
    
    @objc
    public func currentMapInfo(_ callback: @escaping ([String : Any]?) -> ()) {
        self.callApp2Web("currentMapInfo", value: [], callback: { (value: String?) -> () in
            if let value = value, value != "" {
                callback(self.toDictionary(value))
            } else {
                callback(nil)
            }
        })
    }
    
    @objc
    public func mapInfo(_ sourceID: String, callback: @escaping ([String: Any]?) -> ()) {
        self.callApp2Web("mapInfo", value: [sourceID], callback: { (value: String?) -> () in
            if let value = value, value != "" {
                callback(self.toDictionary(value))
            } else {
                callback(nil)
            }
        })
    }
    
}

